frontend/public/components/marketplace/kubernetes-marketplace-items.jsx

```diff
@@ -0,0 +1,298 @@
+      // Apply current filter state to new filters
+      updatedState.filters = copyActiveFilters(newFilters, filters);
+    }
+
+    // If only filters have changed, update filter items and filter counts
+    if (updatedState.filters || filters !== prevState.filters) {
+      Object.assign(updatedState, this.getFilterItemsAndCounts(updatedState.filters || filters));
+    }
+
+    this.setState(updatedState);
+  }
+
+  getFilterItemsAndCounts(filters) {
+    const itemsAndCounts = {
+      filteredItems: {},
+      filterCounts: {},
+    };
+
+    // Filter items by each filter category individually
+    // Example:
+    //  filteredByCategory = {
+    //    provider: [/*array of items filtered by provider*/],
+    //    healthIndex: [/*array of items filtered by healthIndex*/],
+    //  };
+    const filteredByCategory = this.filterByCategory(filters);
+
+    // The set intersection of the above arrays
+    // results in an array of items filtered by all categories
+    const filteredItems = _.intersection(..._.toArray(filteredByCategory));
+    itemsAndCounts.filteredItems = _.sortBy(this.filterByName(filteredItems, filters), 'name');
+
+    // Get counts for filtered items in each category
+    itemsAndCounts.filterCounts = this.getFilterCountsFromCategories(filters, filteredByCategory);
+
+    return itemsAndCounts;
+  }
+
+  filterByCategory(filters) {
+    const {items} = this.props;
+
+    // Filters as follows:
+    //   Active filters in the same category result in the union of items shown
+    //   Active filters in separate categories result in the intersection
+    const filteredByCategory = _.reduce(filters, (filtered, category, key) => {
+      if (key === 'name') {
+        return filtered;
+      }
+
+      // Apply each filter individually to items
+      let categoryActive = false;
+      const categoryItems = _.reduce(category, (accum, elem) => {
+        if (elem.active) {
+          categoryActive = true;
+          accum.push(_.filter(items, [ key, elem.value ]));
+        }
+        return accum;
+      }, []);
+
+      // Filter items only if at least one filter in the category is active
+      // Run union on items only if filter category is active
+      filtered[key] = categoryActive ? _.union(...categoryItems) : items;
+      return filtered;
+    }, {});
+
+    // Returns items filtered by each category
+    return filteredByCategory;
+  }
+
+  filterByName(items, filters) {
+    const { name } = filters;
+    if (name.active) {
+      const filterString = name.value.toLowerCase();
+      return _.filter(items, item => item.name.toLowerCase().includes(filterString));
+    }
+    return items;
+  }
+
+  getFilterCountsFromCategories(filters, itemsFilteredByCategory) {
+    // Takes in filters and object mapping filter property to item array
+    // Returns filterCounts object
+    const { items } = this.props;
+
+    return _.reduce(filters, (counts, category, key) => {
```

Comment:
@SamiSousa Sorry, I'm having trouble following the logic here. Can you give an example of what this is doing?

Reply:
The count next to each filter needs to equal the number of new items that would appear if the filter were selected.

If there were multiple filter groups (such as `healthIndex` and `provider`), selecting a filter under `provider` should filter out items and change the counts under `healthIndex` but not changing the counts for the other filters under `provider`. This is because filters are applied as an intersection of items across filter groups, but as a union for filters within the same filter group.

What I've done is filter the items by each filter group separately, then find the intersection for all groups except the group being counted.

Reply:
I feel that there should be a more efficient and straightforward way of counting, so if you have any ideas I'd love to discuss it further.

Reply:
Once this and #713 are merged, I'd like to try to consolidate the filtering and categorization of catalog and marketplace into a single utility. I would like to do that as part of https://jira.coreos.com/browse/CONSOLE-764

Reply:
I think part of my confusion is that we're using the word "category" in different ways in each catalog. Categories in the Home -> Catalog refer to groups the items go in, but here it is referring to the different filters.

Reply:
It shouldn't. This page will also have categories.

Reply:
Renamed to "groups" when referring to filters